src/desktop/widgets/groupedtoolbutton.cpp

```diff
@@ -63,78 +106,69 @@ void GroupedToolButton::setColorSwatch(const QColor &c)
+		resetStyle();
 	}
+
+	QToolButton::changeEvent(event);
+}
+
+void GroupedToolButton::paintEvent(QPaintEvent *)
+{
 	QStylePainter painter(this);
 	QStyleOptionToolButton opt;
 	initStyleOption(&opt);
 
+	painter.drawComplexControl(QStyle::CC_ToolButton, opt);
+
 	// Color swatch (if set)
 	if(m_colorSwatch.isValid()) {
-		const int swatchH = 5;
 		const QRect swatchRect = QRect(opt.rect.x(), opt.rect.bottom()-swatchH, opt.rect.width(), swatchH);
 		painter.fillRect(swatchRect, m_colorSwatch);
 		opt.rect.setHeight(opt.rect.height() - swatchH);
 	}
 
-	QStyleOptionToolButton panelOpt = opt;
-
-	// Panel
-	QRect& panelRect = panelOpt.rect;
-	switch (mGroupPosition) {
-	case GroupLeft:
-		panelRect.setWidth(panelRect.width() * 2);
-		break;
-	case GroupCenter:
-		panelRect.setLeft(panelRect.left() - panelRect.width());
-		panelRect.setWidth(panelRect.width() * 3);
-		break;
-	case GroupRight:
-		panelRect.setLeft(panelRect.left() - panelRect.width());
-		break;
-	case NotGrouped:
-		Q_ASSERT(0);
-	}
-	painter.drawPrimitive(QStyle::PE_PanelButtonTool, panelOpt);
-
-	// Separator
-	const int y1 = opt.rect.top() + 6;
-	const int y2 = opt.rect.bottom() - 6;
-	if (mGroupPosition & GroupRight) {
+	// Separators
+	const int y1 = opt.rect.top() + 5;
+	const int y2 = opt.rect.bottom() - 5;
+	if (m_groupPosition & GroupRight) {
 		const int x = opt.rect.left();
-		painter.setPen(opt.palette.color(QPalette::Light));
+		painter.setPen(opt.palette.light().color());
 		painter.drawLine(x, y1, x, y2);
 	}
-	if (mGroupPosition & GroupLeft) {
+	if (m_groupPosition & GroupLeft) {
 		const int x = opt.rect.right();
-		painter.setPen(opt.palette.color(QPalette::Mid));
+		painter.setPen(opt.palette.color(QPalette::Dark));
 		painter.drawLine(x, y1, x, y2);
 	}
+}
 
-	const bool showDropdownArrow = menu() != nullptr && !text().isEmpty();
-
-	QRect textRect = opt.rect;
-	QRect arrowRect;
-
-	if(showDropdownArrow) {
-		arrowRect = QRect(textRect.right() - 20, textRect.y(), 20, textRect.height());
-		textRect.setWidth(textRect.width() - arrowRect.width());
-	}
-
-	// Text
-	opt.rect = textRect;
-	painter.drawControl(QStyle::CE_ToolButtonLabel, opt);
-
-	// Dropdown arrow
-	if(showDropdownArrow) {
-		opt.rect = arrowRect;
-		painter.drawPrimitive(QStyle::PE_IndicatorArrowDown, opt);
+void GroupedToolButton::resetStyle()
+{
+	if (!m_resettingStyle) {
+		// Calling `setStyle` emits `QEvent::StyleChange`
+		QScopedValueRollback guard(m_resettingStyle, true);
+		const auto *parentStyle = parentWidget()
+			? parentWidget()->style()
+			: QApplication::style();
+
+		// `QProxyStyle` takes ownership of the base style, but `setStyle` does
+		// not take ownership of whatever style is passed to it, so we need to
+		// track and delete our own proxy styles (but not whatever may be
+		// returned that is inherited since that is owned by something else)
+		auto *oldStyle = qobject_cast<GroupedToolButtonStyle *>(style());
+		setStyle(new GroupedToolButtonStyle(parentStyle->name()));
```

Comment:
This leaks in exactly one case, I guess because Qt isn't calling the destructor or something. Setting the destructor `virtual` doesn't help, but using Qt-sanctioned ownership does. Qt5 also doesn't have `QStyle::name`, but `QObject.:objectName` works for me (probably something to move into qguicompat).

```suggestion
#	if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
		QString styleName = parentStyle->name();
#else
		QString styleName = parentStyle->objectName();
#endif
		auto *newStyle = new GroupedToolButtonStyle(styleName);
		newStyle->setParent(this);
```

And also removing the explicit destructor, since it's no longer needed when Qt is doing it.

Reply:
I’ve revised this so it should be OK now. I ran a leak check and this didn’t look like it was leaking any more, but about a dozen things in QMacStyle are 🙃